File: KannadaNudiAndroid/app/src/main/java/com/kannadanudi/keyboard/prediction/Trie.kt
package com.kannadanudi.keyboard.prediction

class Trie {
    private val root = TrieNode()

    fun insert(word: String) {
        var node = root
        for (char in word) {
            node = node.children.computeIfAbsent(char) { TrieNode() }
        }
        node.isEndOfWord = true
    }

    fun searchPrefix(prefix: String, limit: Int = 10): List<String> {
        val results = ArrayList<String>()
        var node = root

        // Navigate to the end of the prefix
        for (char in prefix) {
            node = node.children[char] ?: return emptyList()
        }

        // Collect words from this node
        collectWords(node, prefix, results, limit)
        return results
    }

    private fun collectWords(node: TrieNode, currentPrefix: String, results: MutableList<String>, limit: Int) {
        if (results.size >= limit) return

        if (node.isEndOfWord) {
            results.add(currentPrefix)
        }

        // To keep it simple and somewhat deterministic, we iterate through keys.
        // For better quality, we might want to prioritize based on frequency if available,
        // but the wordlist is flat.
        // Sort keys to ensure alphabetical order since we switched to HashMap
        val sortedKeys = node.children.keys.sorted()
        for (char in sortedKeys) {
            val childNode = node.children[char]!!
            if (results.size >= limit) return
            collectWords(childNode, currentPrefix + char, results, limit)
        }
    }
}

File: KannadaNudi-iOS/KannadaKeyboard/Logic/Trie.swift
import Foundation

class TrieNode {
    var children: [Character: TrieNode] = [:]
    var isEndOfWord: Bool = false
}

class Trie {
    private let root = TrieNode()

    func insert(_ word: String) {
        var node = root
        for char in word {
            if node.children[char] == nil {
                node.children[char] = TrieNode()
            }
            node = node.children[char]!
        }
        node.isEndOfWord = true
    }

    func searchPrefix(_ prefix: String, limit: Int = 10) -> [String] {
        var results = [String]()
        var node = root

        // Navigate to the end of the prefix
        for char in prefix {
            guard let child = node.children[char] else {
                return []
            }
            node = child
        }

        collectWords(node: node, currentPrefix: prefix, results: &results, limit: limit)
        return results
    }

    private func collectWords(node: TrieNode, currentPrefix: String, results: inout [String], limit: Int) {
        if results.count >= limit { return }

        if node.isEndOfWord {
            results.append(currentPrefix)
        }

        // Sort keys to ensure deterministic order (alphabetical)
        let sortedKeys = node.children.keys.sorted()

        for char in sortedKeys {
            if results.count >= limit { return }
            if let childNode = node.children[char] {
                collectWords(node: childNode, currentPrefix: currentPrefix + String(char), results: &results, limit: limit)
            }
        }
    }
}
